### ISSUE TITLE
Reselect Dart2 xxx  is not a subtype of type '(R1) => T' of 'func'

### ISSUE THREAD

**Issue by neckaros** [brianegan/reselect_dart#2]
Hello,

i have an issue where everytime i try to use reselect with a StoreConnector i get the following error:

```
I/flutter ( 7219): The following assertion was thrown building StoreConnector<AppState, List<LibraryMovie>>:
I/flutter ( 7219): type '(List<LibraryMovie>) => List<LibraryMovie>' is not a subtype of type '(R1) => T' of 'func'
I/flutter ( 7219):
I/flutter ( 7219): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter ( 7219): more information in this error message to help you determine and fix the underlying cause.
I/flutter ( 7219): In either case, please report this assertion by filing a bug on GitHub:
I/flutter ( 7219):   https://github.com/flutter/flutter/issues/new
I/flutter ( 7219):
I/flutter ( 7219): When the exception was thrown, this was the stack:
I/flutter ( 7219): #0      createSelector1 (package:reselect/reselect.dart:235:27)
I/flutter ( 7219): #1      moviesIs4KSelector (file:///C:/xxx/selectors.dart:12:65)
```

But i followed the Readme (adding types only):

Widget connection part:
```
return new StoreConnector<AppState, Iterable<LibraryMovie>>(
        converter: (store) => moviesIs4KSelector(store.state),
        builder: (context, movies) {
          return new GridView.count(
...
```
Selectors:
```
typedef R Func1<A, R>(A a);

final Func1<AppState, MoviesState> moviesStateSelector = (AppState state) => state.moviesState;
final Func1<AppState, Iterable<Library>> librariesSelector = (AppState state) => moviesStateSelector(state).libraries;
final Func1<AppState, List<LibraryMovie>> moviesSelector = (AppState state) => librariesSelector(state).expand((f) => f.folders).expand((m) => m.movies).toList();

final  Func1<AppState, List<LibraryMovie>> moviesIs4KSelector = createSelector1<AppState, List<LibraryMovie>, List<LibraryMovie>>(
    moviesSelector,
    (products) => products
    .where((m) => m.video== null ? false : m.video.quality == "1080p").toList(),
);
```

If however i don't use reselect but a simple selector everything works fine:
`final Func1<AppState, Iterable<LibraryMovie>> moviesIs4KSelector = (AppState state) => moviesSelector(state).where((m) => m.video== null ? false : m.video.quality == "1080p");`

Any idea where this is coming from?

thanks in advance for your help!

**Comment by Paul-Todd** [comment]
I am seeing a very similar issue also with Dart 2.

The following assertion was thrown building StoreConnector<AppState, AppointmentListViewModel>:
type '(List<AppointmentModel>, List<ClientModel>) => List<ApppointmentWithUserModel>' is not a
subtype of type '(R1, R2) => T' of 'func'

The code looks like

``` dart
final Selector<AppState, List<AppointmentModel>> appointmentsSelector = (AppState state) => state.appointments;

final Selector<AppState, List<ClientModel>> clientsSelector = (AppState state) => state.clients;

final Selector<AppState, List<ApppointmentWithUserModel>> appointmentsWithUserSelector =
  createSelector2(appointmentsSelector, clientsSelector, (List<AppointmentModel> appointments, List<ClientModel> clients) {
      final List<ApppointmentWithUserModel> list = new List<ApppointmentWithUserModel>(appointments.length);

      return list;
    }
  );
```

this is called as follows:
```dart
final List<ApppointmentWithUserModel> appointments = appointmentsWithUserSelector(store.state);
```

I can call appointmentsSelector and clientsSelector individually and they work fine, but combining the two raises the above error.

Flutter doctor (-v)
[✓] Flutter (Channel beta, v0.2.3, on Mac OS X 10.13.4 17E199, locale en)
    • Flutter version 0.2.3 at /Users/i063124/Documents/flutter
    • Framework revision 5a58b36e36 (5 weeks ago), 2018-03-13 13:20:13 -0700
    • Engine revision e61bb9ac3a
    • Dart version 2.0.0-dev.35.flutter-290c576264

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/i063124/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = /Users/i063124/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3, Build version 9E145
    • ios-deploy 1.9.2
    • CocoaPods version 1.4.0

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] VS Code (version 1.22.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Dart Code extension version 2.11.2

[✓] Connected devices (1 available)
    • iPhone 8 • F9E245BC-7159-47FB-BF7C-FC589D822121 • ios • iOS 11.3 (simulator)

• No issues found!

flutter analyze
Analyzing /Users/i063124/Documents/fitman_pro...
No issues found!
Ran in 11.5s

Is this a Dart 2 issue?

**Comment by brianegan** [comment]
Hey all, sorry about that trouble! This appears to be a Dart 2 issue. I'll try to patch it up in the next couple of days.

**Comment by brianegan** [comment]
And, actually just made some time to hopefully fix it :) Could you please try the following version of the lib and see if it solves the problems you've encountered:

```yaml
dependencies:
  reselect:
    git:
      url: git://github.com/brianegan/reselect_dart.git
      ref: dart-2
```

If it works out, I'll publish it as a new release.

**Comment by Paul-Todd** [comment]
Seems to have fixed it for me - thanks!!

**Comment by neckaros** [comment]
I still have the issue with this dumbed down selector:
```
final  Func1<AppState, Iterable<LibraryMovie>> distinctMoviesSelector = createSelector1<AppState, Iterable<LibraryMovie>, Iterable<LibraryMovie>>(
    moviesSelector,
    (movies) => movies
);
```
```
I/flutter (11360): The following assertion was thrown building StoreConnector<AppState, Iterable<LibraryMovie>>:
I/flutter (11360): type '(Iterable<LibraryMovie>) => Iterable<LibraryMovie>' is not a subtype of type '(R1) => T' of
I/flutter (11360): 'func'
I/flutter (11360):
I/flutter (11360): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter (11360): more information in this error message to help you determine and fix the underlying cause.
I/flutter (11360): In either case, please report this assertion by filing a bug on GitHub:
I/flutter (11360):   https://github.com/flutter/flutter/issues/new
I/flutter (11360):
I/flutter (11360): When the exception was thrown, this was the stack:
I/flutter (11360): #0      createSelector1 (package:reselect/reselect.dart:235:27)
I/flutter (11360): #1      distinctMoviesSelector (file:///C:/Devs/rs-flutter/redseat/lib/selectors.dart:23:73)
```
Package being: C:\Devs\flutter\.pub-cache\git\reselect_dart-349bf201a2e3bf54fb2fc5579ce440516b4aea14\lib\reselect.dart

**Comment by brianegan** [comment]
@neckaros Hrm, dang... Would you mind providing a bit more code or be willing to share your project? I'm no longer seeing that error with my test cases, but it's totally possible that I'm missing a test case somewhere.

The only problem I can spot is that you're returning a `Func1` instead of a `Selector`. Does this work? 

```dart
final  Selector<Iterable<LibraryMovie>, AppState> distinctMoviesSelector = createSelector1<AppState, Iterable<LibraryMovie>, Iterable<LibraryMovie>>(
    moviesSelector,
    (movies) => movies
);
```

Update: After reading that more carefully and based on Paul-Todd's comment, you can just memoize this function rather than having a converter return the movies:

`final distinctMoviesSelector = memo1(moviesSelector)`

That said, if `moviesSelector` is just reading the value from the store directly (e.g. the function looks like `moviesSelector(AppState state) => state.movies`), probs don't need to memoize it at all, as that access should be extremely fast and memoizing will just consume memory

**Comment by Paul-Todd** [comment]
Not too sure if its relevant but I would also avoid the typedef as well - typedef R Func1<A, R>(A a);

Also the createSelector1 function is taking 3 parameters instead of 2 - is this correct?

As input it should take a selector which is Iterable<LibraryMovie> and not two of them.
```
final  Func1<AppState, Iterable<LibraryMovie>> distinctMoviesSelector = createSelector1<AppState, Iterable<LibraryMovie>, Iterable<LibraryMovie>>(

```

**Comment by brianegan** [comment]
Yep, it requires 3 types since you've got 2 functions! The flow works like this: 

Your `selector` function goes from `S -> R1`, the `combiner` function goes from `R1 -> T`. Therefore, you've got 3 types: `S, R1, T`.

If you don't need a `cominber` function, you could consider simply memoizing your function like so:

`final memoizedSelector = memo1<S, T>(selector)`

`memo1` comes from the `memoize` package that reselect depends on.

Selectors are helpful for a slightly different case: Say you have a base selector that goes `AppState -> List<Todo>`. Then, you want to convert the `List<Todo>` into a `bool`, such as `allTodosComplete`. This is where you'd need the extra function (the `combiner`) and a 3rd type: `AppState -> List<Todo> -> bool`

**Comment by brianegan** [comment]
@neckaros Were you able to play with this any more and find the issue? I'd like to release a new version for Dart 2, and I think the branch I've got up should do the trick, but want to be sure what I've done fixes the problem :)

**Comment by neckaros** [comment]
I will give it another try and get back to you
i should use:
```
git:
      url: git://github.com/brianegan/reselect_dart.git
      ref: dart-2
```
?

**Comment by Lisenish** [comment]
@brianegan Hi! I'm migrating the project to dart 2.0 now and encountered the same problem. I see there is a fix merged in master but why didn't you publish this to the pub? I have to use an override to git branch because of this, could you publish fix?

**Comment by brianegan** [comment]
Oh dang, that's my bad -- totally thought I'd pushed it! Now published as 0.3.0. Thanks for the heads up.
